OPHD/States/MapViewState.cpp

```diff
@@ -725,18 +725,26 @@ void MapViewState::onSystemMenu()
 
 
 /**
- * Changes the current view depth.
+ * Changes the current view depth. If a value is given for the 2nd parameter, the first will be ignored.
  */
-void MapViewState::changeViewDepth(int depth)
+void MapViewState::changeViewDepth(int depth, std::optional<MapOffset> direction)
```

Comment:
This seems to complicate the semantics a bit. Though perhaps the semantics of the view change code are already a bit more complicated than I'd like.

I'm wondering if we need a separate method `onDepthChange`, split off from `changeViewDepth`, which can handle the side effects of changing the depth. Right now we have a couple ways of changing the depth, either by offsetting it from the old value, or by setting it directly. Both might potentially clamp the values to the map dimensions. Both might (potentially) need to run side effects of a depth change. By moving the side effects to a separate function, it would be possible to repeat them when either offsetting or setting directly.

I noticed the `changePoliceOverlayDepth` function expects to know both the `oldDepth` and the `newDepth`. I suppose an `onDepthChange` method could take both values when running the event. Though I wonder if `changePoliceOverlayDepth` really needs both values. I haven't looked at it in ages, and it seems odd to me that it should need to know the old depth.

Reply:
Another option might be to only use one method of changing the depth. In this case I would suggest setting it directly, as that appears to be the only that clamps the values. The key press event handlers could calculate what the value would be for a PageUp/Down press and pass that into the existing changeViewDepth() function. 

Something like the for the PageUp event for example:
`int depth = mMapView->currentDepth() + MapOffsetUp.z;`
`changeViewDepth(depth);`

Reply:
Values are always clamped. That's especially important for relative offsets.

For relative offsets, when calling `MapView::moveView`, it offsets the values and calls `MapView::mapViewLocation`, which handles the clamping. The clamping is direct for `x` and `y`. For the `z` value, there's a further call to `MapView::currentDepth` which does the clamping.

For directly setting the depth, `MapViewState::changeViewDepth` makes a direct call to `MapView::currentDepth`, same as above, to clamp the `z` value.

----

Perhaps the issue here is there is a varying level of abstraction. For offsetting values, there are direct calls to the `MapView` object which handles the offsetting and the clamping. For setting a `z` value directly, there is instead a call on the `MapViewState` object, which then call `MapView` functions. We should perhaps always be calling a method on the same object if we want to maintain a consistent level of abstraction.

The `MapView` class doesn't know about `MapViewState`, so it can't trigger events on `MapViewState` for a changed `z` value. At least, it can't notify it directly. Maybe with a `Signal` or `Delegate` it could trigger something, but I'd like to avoid that kind of complexity.

Maybe what we need is to route both types of updates through methods on `MapViewState`, so it can intercept changes to `z` values and run a method such as `onDepthChange` suggested above. In that way, it can be a simple call when a `z` value change is detected. Perhaps we need a `MapViewState::moveView` method, which delegates to `MapView::moveView`, but also includes a check for changed `z` values.

I think I would check for changes to `z` by recording the old value before doing the update, and checking if it changed after the update. That way, if there's an attempt to change the `z` value, but clamping prevents any real change, we can avoid running code for the side effects. That would be more accurate than checking if there's an attempt to change the `z` value, which might be prevented by clamping.

Reply:
I'm maybe a little worried my thoughts above were a bit too abstract. Maybe I should outline kind of what I'm thinking in terms of refactoring I might attempt.

1. Split `MapViewState::changeViewDepth` so side effects are handled by a new method `MapViewState::onChangeDepth`.
2. Add a new method `MapViewState::moveView`, which delegates to `MapView::moveView`
3. Update calls in `MapViewState` to use `MapViewState::moveView` instead of directly calling `MapView::moveView`
4. Update `MapViewState::moveView` to detect `z` changes, and call the new `onChangeDepth` method. This would fix the bug.
5. Optional: Update `MapViewState::changeViewDepth` to detect `z` changes (which might have been prevented from clamping), and only run the `onChangeDepth` method when there's an actual real `z` value change. This would improve the efficiency of already working code.

Reply:
I'm with Dan on this one. I've recently been reading the book Five Lines of Code and the refactoring suggestions match what Dan outlines above.

Reply:
Those changes make sense. Also, thanks for the new book suggestion. I'll have to take a look at that.